doc reference: add reference for "pgroonga_condition()" function (#321)
pgroonga/pgroonga.github.io: reference/functions/pgroonga-condition.md

---
title: pgroonga_condition function
upper_level: ../
---

# `pgroonga_condition()` function

## Summary

`pgroonga_condition()`関数は`pgroonga_condition`型の値を返します。
関数名と型名が同じですが別物です。
`pgroonga_condition`型は`pgroonga_full_text_search_condition`型や`pgroonga_full_text_search_condition_with_scorers`型のように複雑な条件式を表現します。

`pgroonga_condition()`関数は`pgroonga_condition`型の値を作るための便利関数です。
特定の属性値のみを指定して`pgroonga_condition`型の値を作れます。

`pgroonga_full_text_search_condition`型や`pgroonga_full_text_search_condition_with_scorers`型では、このような便利関数がなかったため
必ず全ての属性値を指定して値を作る必要がありました。

> 

Author: `attribute value`で良いです。
`pgroonga_condition` type はユーザーが独自で定義する型（int型とか、char型とかデータ型の一種）なのですが、それを作る`CREATE TYPE`というコマンドの説明には属性名のことを attribute name と書いているので、ここでもattributeを使うのが良いと思います。

https://www.postgresql.jp/document/15/html/sql-createtype.html

したがって、不要な属性値があっても、`pgroonga_full_text_search_condition`型や`pgroonga_full_text_search_condition_with_scorers`型では、
以下のように不要な属性値には`NULL`を指定する必要がありました。

> 

Author: 不要な属性値とは、「値を作るのに不要な属性値」であってます。

不要な属性値なんですが、属性値には何かしら値を入れないと行けないので、無効な値であることを表す`NULL`を入れる必要がある。ということを言いたいです。

> 

Reviewer: Therefore, you need to designate `NULL` for disused attribute value as follows when `pgroonga_full_text_search_condition` type and `pgroonga_full_text_search_condition_with_scorers` type are used to avoid making all the values.

それであれば不必要の　unnecessaryより、不使用的な意味にのdisusedのほうが良さそうなのでそちらに修正しました。

よろしくお願いいたします。

```
title &@~ ('keyword', NULL, 'index_name')::pgroonga_full_text_search_condition
title &@~ ('keyword', ARRAY[1,1,1,5,0], NULL, 'index_name')::pgroonga_full_text_search_condition_with_scorers
```

型を指定して直接値を作る従来の方法では後方互換性を維持したまま新しい属性を作ることができませんでした。
そのため、新しい属性を追加するたびに`pgroonga_full_text_search_condition_with_XXX`というような新しい型を追加する必要がありました。
たとえば、`pgroogna_full_text_search_condition_with_scorers`型はそのために追加された型です。

`pgroonga_full_text_search_condition`型と`pgroonga_full_text_search_condition_with_scorers`型
の違いは`scorers`が存在するかどうかですが、`pgroonga_full_text_search_condition`型に`scorers`を追加してしまうと、
`scorers`を使わないユーザーも新たに`NULL`を挿入して`pgroonga_full_text_search_condition`型の値を作らなければなりません。

しかし、`pgroonga_condition`型の値を作るための便利関数`pgroonga_condition()`関数を導入することにより後方互換性を維持したまま
`pgroonga_condition`型に新しい属性を追加できます。
`pgroonga_condition()`関数が非互換を吸収してくれるからです。

> 

Reviewer: 32-38

The difference between `pgroonga_full_text_search_condition` type and `pgroonga_full_text_search_condition_with_scorers` type is whether `scorers` exist or not. If `scorers` is added to `pgroonga_full_text_search_condition` type, every users are required to insert new 'NULL' to make `pgroonga_full_text_search_condition` type regardless of `scorers` usage.

However, installing `pgroonga_condition()` function to make new `pgroonga_condition` type value let a new attribution to be added while keeping backward compatibility.
It is because `pgroonga_condition()` function absorb incompatibility.

以下のように、`pgroonga_condition()`関数は不要な属性値を省略できるため、新たに属性値が追加されても既存の書き方を維持できます。
(下記の例では、`weights`、`scorers`、`schema_name`、`column_name`を省略しています。
属性値の詳細については、後述の「構文」で記載します。ここでは、不要な属性値が省略できることに注目してください。)

> 

Reviewer: 40-42

`pgroonga_condition()` function let current writing style when a new attribute value is added because the function can leave out unnecessary attribution value as following sample.
( In the following sample, `weights`、`scorers`、`schema_name`、`column_name` are left out. The details of attribute values would be noted in next ##Sytax. Here, point is that possibility of leaving out unnecessary attribute values.)

```
title &@~ pgroonga_condition('keyword', index_name => 'index_name')
```

`pgroonga_condition()`関数では、上記のように属性値を省略できますが、代わりに、「index_name => 'index name'」のように
キーワード引数のような記載が必要になることに注意してください。

上記の例では、キーワード引数のような書き方をしている属性値とそうでない属性値があります。
どのように書き分けるかについては、後述の「構文」で記載します。
ここでは、従来とは異なる書き方が必要になることがあるという点に注目してください。

> 

Reviewer: 48-53

Please note that while using `pgroonga_condition()` functionm you can leaving out attribute values instead you need to describe comment like key word argument such as "index_name => 'index name'".

In the above sample, there are mix of attribute values which is like key word argument or not.
How to separate writing is going to be explained in next ## Syntax.
The point here is there is need of different writing from the current.

---
title: pgroonga_condition function
upper_level: ../
---

# `pgroonga_condition()` function

## Summary

> `pgroonga_condition()` function return `pgroonga_condition` type value.

Reviewer 2: ```suggestion
`pgroonga_condition()` function returns `pgroonga_condition` type value.
```

The function and the type have same name, but they are two different things.

> `pgroonga_condition` type represent complicated conditional expressions, such as `pgroonga_full_text_search_condition` type and `pgroonga_full_text_search_condition_with_scorers` type. 

Reviewer 2: ```suggestion
`pgroonga_condition` type represents complicated conditional expressions, such as `pgroonga_full_text_search_condition` type and `pgroonga_full_text_search_condition_with_scorers` type. 
```

`pgroonga_condition()` function is a useful function to make the `pgroonga_condition` type value.

> It allows to make the `pgroonga_condition` type value by designating the specific attribute value.

Reviewer 2: ```suggestion
It allows to make the `pgroonga_condition` type value by designating the specific attribute values.
```

designateって初めて見ました。私はspecifyを使っていました。

There were not this kind of useful functions for `pgroonga_full_text_search_condition` type and `pgroonga_full_text_search_condition_with_scorers` type, so designating all attribute values was necessary to make the value.

Therefore, you need to designate `NULL` for disused attribute value as follows when `pgroonga_full_text_search_condition` type and `pgroonga_full_text_search_condition_with_scorers` type are used to avoid making all the values.

```
title &@~ ('keyword', NULL, 'index_name')::pgroonga_full_text_search_condition
title &@~ ('keyword', ARRAY[1,1,1,5,0], NULL, 'index_name')::pgroonga_full_text_search_condition_with_scorers
```

It was not possible for existing value creation methods to make new attribute value while keeping backward compatibility.

> Thus, it was necessary to add a new type every time a new attribute value is added, such as `pgroonga_full_text_search_condition_with_XXX` type.

Reviewer 2: ```suggestion
Thus, it was necessary to add a new type every time when a new attribute value is added, such as `pgroonga_full_text_search_condition_with_XXX` type.
```

> For example, `pgroogna_full_text_search_condition_with_scorers` type was added because of the added new attribution.

Reviewer 2: ```suggestion
For example, `pgroogna_full_text_search_condition_with_scorers` type was added because of the added new attribute.
```

> The difference between `pgroonga_full_text_search_condition` type and `pgroonga_full_text_search_condition_with_scorers` type is whether `scorers` exist or not. If `scorers` is added to `pgroonga_full_text_search_condition` type, every users are required to insert new 'NULL' to make `pgroonga_full_text_search_condition` type regardless of `scorers` usage.

Reviewer 2: ```suggestion
The difference between `pgroonga_full_text_search_condition` type and `pgroonga_full_text_search_condition_with_scorers` type is whether `scorers` exist or not. If `scorers` is added to `pgroonga_full_text_search_condition` type, every users are required to insert new `NULL` to make `pgroonga_full_text_search_condition` type regardless of `scorers` usage.
```

> However, installing `pgroonga_condition()` function to make new `pgroonga_condition` type value let a new attribution to be added while keeping backward compatibility.

Reviewer 2: ```suggestion
However, installing `pgroonga_condition()` function to make new `pgroonga_condition` type value let a new attribute to be added while keeping backward compatibility.
```

> It is because `pgroonga_condition()` function absorb incompatibility.

Reviewer 2: ```suggestion
It is because `pgroonga_condition()` function absorbs incompatibility.
```

absorbは初めて見ました。私ならhideとか使っちゃいそう。

> `pgroonga_condition()` function let current writing style when a new attribute value is added because the function can leave out unnecessary attribution value as following sample.

Reviewer 2: ```suggestion
`pgroonga_condition()` function lets current writing style when a new attribute value is added because the function can leave out unnecessary attribute value as following sample.
```

> ( In the following sample, `weights`、`scorers`、`schema_name`、`column_name` are left out. The details of attribute values would be noted in next "Sytax". Here, point is that possibility of leaving out unnecessary attribute values.)

Reviewer 2: ```suggestion
(In the following sample, `weights`, `scorers`, `schema_name` and `column_name` are left out. The details of attribute values would be noted in next "Syntax". Here, point is that possibility of leaving out unnecessary attribute values.)
```

"Syntax"はリンクにしたほうがいい。

```
title &@~ pgroonga_condition('keyword', index_name => 'index_name')
```

> Please note that while using `pgroonga_condition()` functionm you can leaving out attribute values instead you need to describe comment like key word argument such as "index_name => 'index name'".

Reviewer 2: ```suggestion
Please note that while using `pgroonga_condition()` function you can leave out attribute values instead you need to describe comment like keyword argument such as `index_name => 'index name'`.
```

> In the above sample, there are mix of attribute values which is like key word argument or not.

Reviewer 2: ```suggestion
In the above sample, there are mix of attribute values which is like keyword argument or not.
```

> How to separate writing is going to be explained in next ## Syntax.

Reviewer 2: Syntaxはリンク。

The point here is there is need of different writing from the current.

## Syntax

Here is the syntax of this function:

```
pgroonga_condition pgroonga_condition(keyword,
                                      weights,
                                      scorers,
                                      schema_name,
                                      index_name,
                                      column_name)
```

`keyword`は検索キーワードです。`text`型です。

> `weights`はそれぞれの値の重要度です。`int[]`型です。

Reviewer 2: 複数の値が存在するというのが匂わされていないので「それぞれ」で「？」となりそう。

Author: 「`weights`は検索対象のカラムの重要度です。検索対象のカラムは複数の場合もあるので、`int[]`型です。」として、複数あることを明示してみました。

`scorers`はそれぞれの値のスコアーを計算する処理です。`text[]`型です。

`schema_name`はシーケンシャルサーチ実行時に参照するインデックスが属するスキーマです。`text`型です。

`index_name`はシーケンシャルサーチ実行時に参照するインデックスです。`text`型です。

> `column_name`はシーケンシャルサーチ実行時に参照するインデックスが紐付けられている属性です。`text`型です。

Reviewer 2: ```suggestion
`column_name`はシーケンシャルサーチ実行時に参照するインデックス内の対象となる属性です。`text`型です。
```

> `pgroonga_condition()`の引数はすべて省略可能です。そのため、[「`引数名 => 値`」][sql-syntax-calling-funcs-named]という名前付き表記を使うことで特定の引数だけ指定することができます。たとえば、`index_name`だけ指定する場合は`pgroonga_condition(index_name => 'index1')`となります。

Reviewer 2: ```suggestion
`pgroonga_condition()`の引数はすべて省略可能です。そのため、[`引数名 => 値`][sql-syntax-calling-funcs-named]という名前付き表記を使うことで特定の引数だけ指定することができます。たとえば、`index_name`だけ指定する場合は`pgroonga_condition(index_name => 'index1')`となります。
```

ただ、一般的なユースケースでは次の3種類の書き方を覚えておけば十分です。

```
title &@~ pgroonga_condition('keyword', index_name => 'pgroonga_index')
title &@~ pgroonga_condition('keyword', ARRAY[weight1, weight2, ...])

> title &@~ pgroonga_condition('keyword', ARRAY[weight1, weight2, ...], index_name => 'pgroonga_index')

Reviewer 2: `title &@~`いらないかも。

```

> 上の例以外の使い方をする場合のために、「`引数名 => 値`」で記述する必要がある引数とそうでない引数の違いを説明します。

Reviewer 2: ```suggestion
上の例以外の使い方をする場合のために、`引数名 => 値`で記述する必要がある引数とそうでない引数の違いを説明します。
```

結局この後にどういうときに記述する必要があるかがビシッと書かれていないので、ここで「○○のときです」と書いておきませんか？

の違いは`scorers`が存在するかどうかですが、`pgroonga_full_text_search_condition`型に`scorers`を追加してしまうと、
`scorers`を使わないユーザーも新たに`NULL`を挿入して`pgroonga_full_text_search_condition`型の値を作らなければなりません。

しかし、`pgroonga_condition`型の値を作るための便利関数`pgroonga_condition()`関数を導入することにより後方互換性を維持したまま
`pgroonga_condition`型に新しい属性を追加できます。
`pgroonga_condition()`関数が非互換を吸収してくれるからです。

以下のように、`pgroonga_condition()`関数は不要な属性値を省略できるため、新たに属性値が追加されても既存の書き方を維持できます。
(下記の例では、`weights`、`scorers`、`schema_name`、`column_name`を省略しています。
属性値の詳細については、後述の「構文」で記載します。ここでは、不要な属性値が省略できることに注目してください。)

```
title &@~ pgroonga_condition('keyword', index_name => 'index_name')
```

`pgroonga_condition()`関数では、上記のように属性値を省略できますが、代わりに、「index_name => 'index name'」のように
キーワード引数のような記載が必要になることに注意してください。

上記の例では、キーワード引数のような書き方をしている属性値とそうでない属性値があります。
どのように書き分けるかについては、後述の「構文」で記載します。
ここでは、従来とは異なる書き方が必要になることがあるという点に注目してください。

## Syntax

Here is the syntax of this function:

```
pgroonga_condition pgroonga_condition(query,
                                      weights,
                                      scorers,
                                      schema_name,
                                      index_name,
                                      column_name)
```

> * `query`: `text`型の値です。クエリー構文を使っているクエリーです。

Reviewer 2: クエリー構文とは限らないです。`&=`で使うときは完全一致対象です。

Author: ```
`keyword`は検索キーワードです。`text`型です。
```

書き方を他のドキュメントとあわせました。
`query`はクエリー構文を想起させるので、`query`を`keyword`に変更しました。

> * `weights`: `int[]`型の値です。検索対象のカラムの重みを指定します。検索対象のカラムが`text`型や`varchar`型のように配列ではない場合、最初の要素が検索対象のカラムの値の重みになります。

Reviewer 2: > 検索対象のカラムが`text`型や`varchar`型のように配列ではない場合、最初の要素が検索対象のカラムの値の重みになります。

こういう個別の話は各演算子のところで説明しませんか？

Author: ```diff
- * `weights`: `int[]`型の値です。検索対象のカラムの重みを指定します。検索対象のカラムが`text`型や`varchar`型のように配列ではない場合、最初の要素が検索対象のカラムの値の重みになります。
+ * `weights`はそれぞれの値の重要度です。`int[]`型です。
```

書き方を他のドキュメントとあわせました。
また、個別の話を削除しました。

* `scorers`: `text[]`型の値です。検索対象のカラムそれぞれのスコアーを計算するスコアラーを指定します。検索対象のカラムが`text`型や`varchar`型のように配列ではない場合、最初の要素が検索対象のカラムの値のスコアーを計算します。
* `schema_name`: `text`型の値です。シーケンシャルサーチを使った検索が実行された際に参照するインデックスが存在するスキーマを指定します。

  通常のケースでは指定する必要はありません。
  PostgreSQLは、スキーマ未指定の場合`search_path`に登録されているスキーマから該当するインデックスを検索します。
  通常は、`search_path`に存在するスキーマ内に該当するインデックスがあるため`schema_name`を指定しなくても適切なインデックスを参照できます。

>   しかし、 例えばpostgres_fdw を使って外部のPostgreSQLのデータベースへアクセスする場合、`search_path`は`pg_catalog`のみになります。

Reviewer 2: `postgres_fdw`はリンクにしてください。

  このケースでは、`pg_catalog`スキーマ内に参照したいインデックスが存在しない場合、スキーマ未指定では該当のインデックスを発見できません。
  このように、`search_path`に登録されているスキーマ以外のスキーマに参照したいインデックスがある場合は、`schema_name`で明示的にスキーマを指定することで
  該当のインデックスを発見できます。

* `index_name`: `text`型の値です。シーケンシャルサーチを使った検索が実行された際に参照するインデックスを指定します。

>   通常、PGroongaのインデックスではなくシーケンシャルサーチが実行されると、PGroongaのインデックスに設定されているトークナイザーとノーマライザーを使えません。

Reviewer 2: 「PGroongaのインデックス」と「シーケンシャルサーチ」は同列ではありません。
同列にするなら「インデックスサーチ」と「シーケンシャルサーチ」です。

  シーケンシャルサーチの場合はPGroongaのインデックスを使わないので、トークナイザーとノーマライザーの設定がどこにあるか判断できないためです。
  そのため、シーケンシャルサーチ実行時と、PGroongaのインデックスを使った時で検索結果が変わってしまうことがあります。
  これを防止するため、この引数でPGroongaのインデックスを指定しシーケンシャルサーチ実行時でもトークナイザーとノーマライザーの設定を参照できるようにしています。

* `column_name`: `text`型の値です。シーケンシャルサーチが実行された際に参照するインデックスが紐付けられているカラムを指定します。

>   PGroongaには、インデックスのオプションに`normalizers_mapping`があります。これは以下のように特定のカラムに対して、特定のノーマライザーとそのオプションを指定できるものです。

Reviewer 2: `normalizers_mapping`はリンクにしてください。

カラムじゃなくて属性にしてください。

  その結果、下記のようにシーケンシャルサーチでも「バイオリン」で「ヴァイオリン」がヒットします。

  ```sql
  SELECT *
    FROM memos
   WHERE title &@~ pgroonga_condition('バイオリン',
                                      index_name => 'pgroonga_memos_index',
                                      column_name => 'title');
   id |    title     | content  
  ----+--------------+----------
    1 | ヴァイオリン | content1 
  (1 row)

  EXPLAIN ANALYZE VERBOSE
  SELECT *
    FROM memos
   WHERE title &@~ pgroonga_condition('バイオリン',
                                      index_name => 'pgroonga_memos_index',
                                      column_name => 'title');
                                                 QUERY PLAN                                               
  --------------------------------------------------------------------------------------------------------
   Seq Scan on public.memos  (cost=0.00..2.52 rows=1 width=100) (actual time=1.830..1.834 rows=1 loops=1)
     Output: id, title, content, tag
     Filter: (memos.title &@~ '(バイオリン,,,,pgroonga_memos_index,title)'::pgroonga_condition)
     Rows Removed by Filter: 1
   Planning Time: 1.307 ms
   Execution Time: 1.927 ms
  (6 rows)
  ```

> `pgroonga_condition()`の引数は省略可能ですが、関数のシグネチャーに定義されている引数の位置と異なる場所に記載する場合は、「index_name => 'index name'」のように
> 「引数名 => 値」の形で記述する必要があります。

Reviewer 2: ```suggestion
`pgroonga_condition()`の引数はすべて省略可能です。そのため、`引数名 => 値`という名前付き表記を使うことで特定の引数だけ指定することができます。たとえば、`index_name`だけ指定する場合は`pgroonga_condition(index_name => 'index1')`となります。
```

`引数名 => 値`の書き方はリンク： https://www.postgresql.org/docs/current/sql-syntax-calling-funcs.html#SQL-SYNTAX-CALLING-FUNCS-NAMED

`pgroonga_condition()`関数が非互換を吸収してくれるからです。

以下のように、`pgroonga_condition()`関数は不要な属性値を省略できるため、新たに属性値が追加されても既存の書き方を維持できます。
(下記の例では、`weights`、`scorers`、`schema_name`、`column_name`を省略しています。
属性値の詳細については、後述の「構文」で記載します。ここでは、不要な属性値が省略できることに注目してください。)

```
title &@~ pgroonga_condition('keyword', index_name => 'index_name')
```

`pgroonga_condition()`関数では、上記のように属性値を省略できますが、代わりに、「index_name => 'index name'」のように
キーワード引数のような記載が必要になることに注意してください。

上記の例では、キーワード引数のような書き方をしている属性値とそうでない属性値があります。
どのように書き分けるかについては、後述の「構文」で記載します。
ここでは、従来とは異なる書き方が必要になることがあるという点に注目してください。

## Syntax

Here is the syntax of this function:

```
pgroonga_condition pgroonga_condition(query,
                                      weights,
                                      scorers,
                                      schema_name,
                                      index_name,
                                      column_name)
```

`query`は検索キーワードです。`text`型です。

`weights`はそれぞれの値の重要度です。`int[]`型です。

`scorers`はそれぞれの値のスコアーを計算する処理です。`text[]`型です。

`schema_name`はシーケンシャルサーチ実行時に参照するインデックスが属するスキーマです。`text`型です。

`index_name`はシーケンシャルサーチ実行時に参照するインデックスです。`text`型です。

`column_name`はシーケンシャルサーチ実行時に参照するインデックスが紐付けられている属性です。`text`型です。

> `pgroonga_condition()`の引数はすべて省略可能です。そのため、[「引数名 => 値」][sql-syntax-calling-funcs-named]という名前付き表記を使うことで特定の引数だけ指定することができます。たとえば、`index_name`だけ指定する場合は`pgroonga_condition(index_name => 'index1')`となります。

Author: 文書内の 引数名 => 値 をすべて `引数名 => 値` に置換しました。

> ただ、一般的なユースケースでは下記の3種類の書き方を覚えておけば十分です。

Author: 「下記」を「次の」に書き換えました。

```diff
- ただ、一般的なユースケースでは下記の3種類の書き方を覚えておけば十分です。
+ ただ、一般的なユースケースでは次の3種類の書き方を覚えておけば十分です。
```

> ただ、一般的なユースケースでは下記の3種類の書き方を覚えておけば十分です。

Author: 他の箇所も修正しました。

```diff
- 以下の例は、シーケンシャルサーチが実行されていますが、「バイオリン」で「ヴァイオリン」がヒットしていることが確認できます。
+ 次の例は、シーケンシャルサーチが実行されていますが、「バイオリン」で「ヴァイオリン」がヒットしていることが確認できます。
```

> ただ、一般的なユースケースでは下記の3種類の書き方を覚えておけば十分です。

Author: 文書全体を検索し、その他に「下記」が残っていないことも確認しました。

```
title &@~ pgroonga_condition('query', index_name => 'pgroonga_index')
title &@~ pgroonga_condition('query', ARRAY[weight1, weight2, ...])
title &@~ pgroonga_condition('query', ARRAY[weight1, weight2, ...], index_name => 'pgroonga_index')
```

> 上記以外の使い方をする場合のために、「引数名 => 値」で記述する必要がある引数とそうでない引数の違いを説明します。

Author: 引数名 => 値 を `引数名 => 値` に置換しました。

例えば、下記は引数`weights`、`scorers`、`schema_name`、`column_name`を省略しています。

```
title &@~ pgroonga_condition('query', index_name => 'pgroonga_index')
```

> 引数`weights`と`scorers`と`schema_name`を省略したことで、引数`index_name`の位置は上記の`pgroonga_condition()`の第2引数ですが、

Author: ↑のdiffの通り修正しました。

関数のシグネチャーでは`index_name`は第5引数なので、このケースでは、`index_name`は関数のシグネチャーと位置が異なる引数となります。

> 一方、上記の`pgroonga_condition()`の第1引数にある`query`は関数のシグネチャーでも第1引数なので、関数のシグネチャーと位置が同じ引数となります。

Author: ↑のdiffの通り修正しました。

> 上記から、関数のシグネチャーと同じ位置にある、`query`は、「引数名 => 値」の形で書く必要はなく、値をそのまま記述できますが、

Author: ↑のdiffの通り修正しました。

関数のシグネチャーと違う位置にある、`index_name`は、「引数名 => 値」の形で書く必要があります。

## Usage

Here are sample schema and data:

```sql
CREATE TABLE memos (
  id integer,
  title text,
  content text
);

CREATE INDEX pgroonga_memos_index
          ON memos
       USING pgroonga (title)
        WITH (normalizers='"NormalizerNFKC150(\"unify_katakana_v_sounds\", true)"');

INSERT INTO memos VALUES (1, 'ヴァイオリン', E'Let\'s play violin!');
```

> シーケンシャルサーチ実行時でも検索結果が変化しないように、シーケンシャルサーチ実行時でも
> `title`属性に設定したインデックスを参照できるようにするには、以下のようにします。

Reviewer 2: ```suggestion
インデックスサーチ時はインデックスに指定したオプションで検索結果をカスタマイズできます。このサンプルだと`normalizers='...'`としてオプションを指定しています。一方、普通にシーケンシャルサーチをするとインデックスに指定したオプションを参照できません。シーケンシャルサーチ時にどのインデックスを参照すればよいかという情報がないからです。そのため、オプションを指定している場合はインデックスサーチ時とシーケンシャルサーチ時で検索結果が異なる可能性があります。その問題を回避するためにシーケンシャルサーチ時に参照するインデックスを明示的に指定できます。`index_name => '...'`がそのための引数です。
```

```sql
SELECT *
  FROM memos
 WHERE title &@~ pgroonga_condition('バイオリン',
                                    index_name => 'pgroonga_memos_index');
（結果）
```

`index_name`はシーケンシャルサーチ実行時に参照するインデックスを指定します。

通常、インデックスサーチではなくシーケンシャルサーチが実行されると、PGroongaのインデックスに設定されているトークナイザーとノーマライザーを使えません。
シーケンシャルサーチの場合はPGroongaのインデックスを使わないので、トークナイザーとノーマライザーの設定がどこにあるか判断できないためです。
そのため、シーケンシャルサーチ実行時と、インデックスサーチ実行時で検索結果が変わってしまうことがあります。
上記の例で、もし`index_name`を指定しなければ、シーケンシャルサーチ実行時は「ヴァイオリン」で「バイオリン」はヒットしないので、
インデックスサーチ実行時とシーケンシャルサーチ実行時で検索結果が変わってしまいます。

> これを防止するため、`index_name`でPGroongaのインデックスを指定しシーケンシャルサーチ実行時でもトークナイザーとノーマライザーの設定を参照できるようにしています。

Reviewer 2: あぁ、ここらへんで書いていたのか。

複数のカラムが検索対象で、カラム毎の重みを設定したい場合は以下のようにします。
`pgroonga_condition()`の第2引数の配列で重みを設定しています。

下記例のように、重みを0にすることで対応するカラムを検索対象外にできます。
下記例では、`content`カラムを検索対象外にしています。

```sql
CREATE TABLE memos (
  id integer,
  title text,
  content text
);

CREATE INDEX pgroonga_memos_index
          ON memos
       USING pgroonga (ARRAY[title, content]::text[])
        WITH (normalizers='"NormalizerNFKC150(\"unify_katakana_v_sounds\", true)"');

INSERT INTO memos VALUES (1, 'ヴァイオリン', E'Let\'s play violin!');
INSERT INTO memos VALUES (2, 'チェロ', 'ヴァイオリンだけではなく、チェロも始めました！');

SELECT *
  FROM memos
 WHERE ARRAY[title, content]::text[] &@~ pgroonga_condition('バイオリン',
                                                            ARRAY[1,0],
                                                            index_name => 'pgroonga_momos_index');
(結果)

> ```

Reviewer 2: うーん、こういう話は`&@~`の方に書いた方がよくない？

`&@~`の使い方を知りたい人が知りたい情報ですよね？`pgroonga_condition()`単体だと役に立たないので、ここにたくさん説明を書くよりも、`&@~`とか`pgroonga_condition()`を使ってなにか効果があるやつの方に書いた方が見つけやすくないですか？

Author: `schema_name`や`column_name`などの個別の説明は、Usageを`&@~`の方に移動しました。

        WITH (normalizers_mapping='{
                "title": "NormalizerNFKC150(\"unify_katakana_v_sounds\", true)",
                "content": "NormalizerNFKC150"
              }',
              normalizers='NormalizerAuto');

INSERT INTO memos VALUES (1, 'ヴァイオリン', E'Let\'s play violin!');

SELECT *
  FROM memos
 WHERE title &@~ pgroonga_condition('バイオリン',
                                    index_name => 'pgroonga_memos_index',
                                    column_name => 'title');
```

`column_name`はシーケンシャルサーチ実行時に参照するインデックスが紐付けられている属性を指定します。

PGroongaには、インデックスのオプションに[`normalizers_mapping`][normalizers-mapping]があります。
これは上記の例のように特定の属性に対して、特定のノーマライザーとそのオプションを指定できるものです。

上記の例では、`title`属性に`unify_katakana_v_sounds`が設定されています。
「バイオリン」で「ヴァイオリン」をヒットさせるためには、`unify_katakana_v_sounds`が有効である必要がありますが、
シーケンシャルサーチが実施された場合、PGroongaのインデックスを参照できず`unify_katakana_v_sounds`が効きません。
そこで、以下のように`pgroonga_condition()`の`column_name`で`title`属性を指定することで、`title`属性に
設定されている`unify_katakana_v_sounds`を使えます。

その結果、上記の例のようにシーケンシャルサーチでも「バイオリン」で「ヴァイオリン」をヒットさせることができます。

## See also

* [postgres_fdw][postgres-fdw]
* [normalizers_mapping][normalizers-mapping]
* [名前付け表記][sql-syntax-calling-funcs-named]

> [postgres-fdw]:https://www.postgresql.org/docs/current/postgres-fdw.html

Reviewer 2: 他のページみたいに`{{ site.postgresql_doc_base_url.en }}`とかを使ってください。

Author: `{{ site.postgresql_doc_base_url.en }}`を使うようにしました。

> [normalizers-mapping]:../create-index-using-pgroonga.html

Reviewer 2: `XXX.html#YYY`まで指定して必要な情報をすぐに参照できるようにしませんか？

Author: こちらも直接参照できるようにしました。

---
title: pgroonga_condition function
upper_level: ../
---

# `pgroonga_condition` function

## Summary

この関数は、
`column operator (keyword, weights, index_name)::pgroonga_full_text_search_condition` や
`column operator (keyword, weights, scorers, index_name)::pgroonga_full_text_search_condition_with_scorers` を

> `pgroonga_condition(keyword, weights, index_name => 'index_name')` のように関数を使った形にできます。

Reviewer 2: 関数を使った形にできるかどうかが重要なポイントではないです。
それはただの実現方法であり実装の詳細です。

重要なポイントは必要なオプションだけ指定できるようになることと、後方互換性を壊したり新しい型を増やさずに新しいオプションを追加できるようになることです。
どちらもユーザーにとってうれしいことです。

---
title: pgroonga_condition function
upper_level: ../
---

# `pgroonga_condition` function

## Summary

この関数は、`pgroonga_full_text_search_condition`や`pgroonga_full_text_search_condition_with_scorers`と

> 同等の機能を持ちます。

Reviewer 2: `pgroonga_full_text_search_condition`とかは型なので、そいつらとこの関数が同等の機能を持つというのは違和感があります。

↓みたいにするとかもう少し説明の仕方はありませんか？

`pgroonga_condition()`関数は`pgroonga_condition`型の値を返します。関数名と型名が同じですが別物です。`pgroonga_condition`型は`pgroonga_full_text_search_condition`型・`pgroonga_full_text_search_condition_with_scorers`型のように複雑な条件式を表現します。

> `pgroonga_condition()`では、`pgroonga_full_text_search_condition`や`pgroonga_full_text_search_condition_with_scorers`と異なり、

Reviewer 2: ここも、関数と型を同列に扱うのは避けませんか？
`pgroonga_condition`型もあるので混乱しやすいと思います。

↓みたいにするとかもう少し説明の仕方はありませんか？

`pgroonga_condition()`は`pgroonga_condition`型の値を作るための便利関数です。特定の属性値のみを指定して`pgroonga_condition`型の値を作ることができます。`pgroonga_full_text_search_condition`型と`pgroonga_full_text_search_condition_with_scorers`型にはこのような便利関数がなかったため、必ずすべての属性値を指定して値を作る必要がありました。

型を指定して直接値を作る従来の方法では後方互換性を維持したまま新しい属性を作ることができませんでした。そのため、新しい属性を追加するたびに`pgroonga_full_text_search_condition_with_XXX`というような型を追加する必要がありました。たとえば、`pgroogna_full_text_search_condition_with_scorers`型はそのために追加された型です。

しかし、`pgroonga_condition`型の値を作るための便利関数`pgroonga_condition()`関数を導入することにより後方互換性を維持したまま`pgroonga_condition`型に新しい属性を追加できます。`pgroonga_condition()`関数が非互換を吸収してくれるからです。

不要な引数を省略できます。
不要な引数を省略できることで、後方互換性を壊すことや新しいユーザー定義の型を作ること無く新しいオプションを追加可能になりました。

`pgroonga_full_text_search_condition`や`pgroonga_full_text_search_condition_with_scorers`では、不要な引数は省略できず
以下のように`NULL`を入れる必要がありました。

```
column operator ('keyword', NULL, 'index_name')::pgroonga_full_text_search_condition
column operator ('keyword', ARRAY[1,1,1,5,0], NULL, 'index_name')::pgroonga_full_text_search_condition_with_scorers
```

いままでは、新しいオプションを追加する際に後方互換性を壊さないように新しい型を追加していました。
例えば、`pgroonga_full_text_search_condition`と`pgroonga_full_text_search_condition_with_scorers`
の違いは、`scorers`を指定できるかどうかですが、もし、`pgroonga_full_text_search_condition`に`scorers`を追加してしまうと、
`scorers`を使わないユーザーも新たに`NULL`を挿入しなければなりません。そうしないとシンタックスエラーになってしまいます。

このように、後方互換性を壊すのを避けるため、新しいオプションを追加する際は合わせて新しいユーザー定義の型を追加していましたが
オプションが増えるごとに型を追加するのは煩雑です。

`pgroonga_condition()`を使えば、以下のように不要なオプションを省略できるので、新しいオプションが追加されてもそれが不要であれば省略できるため
既存の書き方をそのまま継続して使えるようになります。
下記の例では、`weights`、`scorers`、`schema_name`、`column_name`を省略しています。
引数の詳細については、後述の「構文」で記載します。ここでは、不要な引数が省略できることに注目してください。

```

> column operator pgroonga_condition('keyword', index_name => 'index_name')

Reviewer 2: `column operator ...`だとどういうことかわかりにくくないですか？

`title &@~ pgroonga_condition(...)`とかもっとわかりやすい例にしませんか？

以下のように不要な属性値には`NULL`を指定する必要がありました。

```
title &@~ ('keyword', NULL, 'index_name')::pgroonga_full_text_search_condition
title &@~ ('keyword', ARRAY[1,1,1,5,0], NULL, 'index_name')::pgroonga_full_text_search_condition_with_scorers
```

型を指定して直接値を作る従来の方法では後方互換性を維持したまま新しい属性を作ることができませんでした。
そのため、新しい属性を追加するたびに`pgroonga_full_text_search_condition_with_XXX`というような新しい型を追加する必要がありました。
たとえば、`pgroogna_full_text_search_condition_with_scorers`型はそのために追加された型です。

`pgroonga_full_text_search_condition`型と`pgroonga_full_text_search_condition_with_scorers`型
の違いは`scorers`が存在するかどうかですが、`pgroonga_full_text_search_condition`型に`scorers`を追加してしまうと、
`scorers`を使わないユーザーも新たに`NULL`を挿入して`pgroonga_full_text_search_condition`型の値を作らなければなりません。

しかし、`pgroonga_condition`型の値を作るための便利関数`pgroonga_condition()`関数を導入することにより後方互換性を維持したまま
`pgroonga_condition`型に新しい属性を追加できます。
`pgroonga_condition()`関数が非互換を吸収してくれるからです。

以下のように、`pgroonga_condition()`関数は不要な属性値を省略できるため、新たに属性値が追加されても既存の書き方を維持できます。
(下記の例では、`weights`、`scorers`、`schema_name`、`column_name`を省略しています。
属性値の詳細については、後述の「構文」で記載します。ここでは、不要な属性値が省略できることに注目してください。)

```
title &@~ pgroonga_condition('keyword', index_name => 'index_name')
```

> `pgroonga_condition()`関数では、上記のように属性値を省略できますが代わりに、「index_name => 'index name'」のように

Reviewer 2: ```suggestion
`pgroonga_condition()`関数では、上記のように属性値を省略できますが、代わりに、「index_name => 'index name'」のように
```
